[Docs] Add automatic operation documentation generation with PyTorch-style structure

samiyadev786 opened the pull request:
## Summary

This PR implements automatic operation discovery and documentation generation from `forge/forge/op/*.py` files, addressing the requirement for automatic discovery of operations similar to Doxygen/PyTorch documentation style.

## Changes

### 🎯 Core Implementation

1. **Automatic Operation Discovery** (`scripts/discover_operations.py`)
   - New module that automatically scans `forge/forge/op/*.py` files
   - Uses Python AST to parse function definitions
   - Extracts function signatures, docstrings, and parameter information
   - Infers operation categories from file names
   - Discovers **78 operations** automatically from the codebase

2. **Enhanced Documentation Generator** (`scripts/generate_ops_docs.py`)
   - Integrated automatic discovery as the primary method
   - Uses actual Python function signatures from source code
   - Improved docstring parsing with better edge case handling
   - Enhanced parameter description formatting (removes redundant type prefixes)
   - Fixed incorrect descriptions (e.g., "Sigmoid" for Abs → "Elementwise absolute value operation")
   - Cleaner return value descriptions
   - Falls back to `ops_data.py` only if automatic discovery fails

### 📝 Generated Documentation

3. **Operation Documentation Pages**
   - Generated 78 operation pages using `forge.op.*` naming convention
   - All pages follow PyTorch-style documentation format
   - Consistent structure with accurate function signatures
   - Each page includes function signature, parameters, returns, and descriptions

4. **Index Page** (`docs/src/operations.md`)
   - Categorized list of all 78 operations
   - Operations grouped by type (Elementwise, Convolution, Pooling, etc.)
   - Short descriptions for each operation
   - Links to individual operation pages

## Features

✅ **Automatic Discovery**: Operations are automatically discovered from Python source files  
✅ **No Manual Entry Required**: Simply define operations with proper docstrings in `forge/forge/op/*.py`  
✅ **Accurate Signatures**: Uses actual Python function signatures from source code  
✅ **Better Formatting**: Cleaner parameter and return descriptions  
✅ **Category Inference**: Automatically categorizes operations based on file location  
✅ **Edge Case Handling**: Handles missing/incorrect docstrings gracefully  
✅ **PyTorch-Style Structure**: Matches PyTorch documentation format with root page and individual pages  

## How It Works

1. **Discovery Phase** (`discover_operations.py`):
   - Scans all Python files in `forge/forge/op/` directory
   - Identifies operation functions (functions starting with uppercase letters)
   - Uses Python AST to extract function signatures, docstrings, and parameters
   - Infers categories from file names (e.g., `eltwise_unary.py` → "Elementwise Operations")

2. **Conversion Phase** (`generate_ops_docs.py`):
   - Converts discovered operations to documentation format
   - Cleans up descriptions and removes redundant type prefixes
   - Handles edge cases (missing docstrings, incorrect descriptions)
   - Infers better descriptions from function names when docstrings are inadequate

3. **Generation Phase** (`generate_ops_docs.py`):
   - Creates individual markdown pages for each operation
   - Generates categorized index page
   - Formats in PyTorch-style documentation format

## Usage

To regenerate documentation after adding new operations:

```bash
python scripts/generate_ops_docs.py
```

New operations are automatically discovered if they:
- Are defined in `forge/forge/op/*.py` files
- Have function names starting with uppercase letters
- Include proper docstrings (preferably NumPy-style with Parameters and Returns sections)

## Documentation Structure

### Index Page (`operations.md`)
- Operations organized by categories:
  - Elementwise Operations
  - Convolution Functions
  - Pooling Functions
  - Linear Functions
  - Reduction Operations
  - Tensor Manipulation
  - Normalization Functions
  - Resize Operations
  - Embedding Functions
  - Memory Operations
  - Creation Operations
  - Other Operations

### Individual Operation Pages
Each page includes:
- **Function Signature**: Extracted from Python source code with type annotations
- **Parameters**: From docstrings with types, descriptions, and default values
- **Returns**: Return value descriptions with types
- **Categories**: Automatically inferred from file names

## Current Coverage

Automatically discovered **78 operations** from the codebase, including:
- Elementwise: Abs, Add, Subtract, Multiply, Divide, Relu, Sigmoid, Tanh, Gelu, etc.
- Convolution: Conv2d, Conv2dTranspose
- Pooling: MaxPool1d, MaxPool2d, AvgPool1d, AvgPool2d
- Linear: Matmul
- Reduction: ReduceSum, ReduceAvg, ReduceMax, Argmax
- Tensor Manipulation: Reshape, Transpose, Broadcast, Pad, etc.
- Normalization: Softmax, LogSoftmax, Layernorm, Batchnorm, Dropout
- And many more...

## Testing

- ✅ Successfully discovers 78 operations from codebase
- ✅ Generates all operation pages correctly
- ✅ Creates categorized index page
- ✅ Handles edge cases (missing docstrings, incorrect descriptions)
- ✅ Function signatures match actual Python code
- ✅ All links working correctly
- ✅ Integrated with mdBook documentation system

This implementation provides:
- ✅ Automatic discovery of operations from `forge/forge/op/*`
- ✅ Structured operation list similar to PyTorch
- ✅ Individual pages for each operation with all details
- ✅ Root page with operations split by types with short descriptions
- ✅ Automatic discovery of new operations (no manual entry needed)

## Related Issue

Closed: #59


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210939

nvukobratTT reviewed (commented):
@samiyadev786 thanks for your contribution! We'll review it asap! :))

@chandrasekaranpradeep, can you do an initial overview of the change? Have in mind that operation documentation should be generated based on existing ops + automatically discover when new are added and generate docs based on them as well. I.e., referncing this and similar files:
- tenstorrent/tt-forge-fe/forge/forge/op/*

samiyadev786 commented:
> I saw doc generation implementation changes from ttir to forge operation. So waiting for you to complete. will review this PR within today or tomorrow

@chandrasekaranpradeep Thanks!

chandrasekaranpradeep reviewed (changes requested):
Hi @samiyadev786 

Sharing some recommendations on improving the operations documentation structure.

Please review the following:

Documentation structure recommendations:
https://github.com/tenstorrent/tt-forge-fe/blob/0619ed66cfdaaf85bcab41a2cbb6de8e58aa9494/docs/DOCUMENTATION_STRUCTURE.md

(Guidelines for enhancing operations documentation and overall structure)

Improved index page example:
https://github.com/tenstorrent/tt-forge-fe/blob/0619ed66cfdaaf85bcab41a2cbb6de8e58aa9494/docs/src/operations/README.md

Improved individual operation example (Resize2d):
https://github.com/tenstorrent/tt-forge-fe/blob/0619ed66cfdaaf85bcab41a2cbb6de8e58aa9494/docs/src/operations/Resize2d.md

Enhancement suggestion:
The operation descriptions, parameters, and return value sections can be strengthened by referencing the actual C++ implementations under forge/csrc/ops. For example, for Resize2d, reviewing the op_resize_2d.cpp file—especially the shape and eval functions—ensures that documentation stays aligned with real behavior. The same approach can be applied across other operations by checking their corresponding implementation files.

@nvukobratTT @pilkicTT @acolicTT
Please share your thoughts on updating the docstrings to better reflect the implementation details. Currently, our Forge operation docstrings are helpful but not yet on par with the documentation quality found in PyTorch, tt-nn ops, or tt-mlir operation docs.

@samiyadev786 May I know the reason why we are having fallback to ttir operation instead of forge operation?

samiyadev786 commented:
Hi @chandrasekaranpradeep! 

I've implemented all the changes you requested:

✅ Removed TTIR fallback entirely - all docs now sourced from forge/forge/op/*.py
✅ Added fail-fast validation - throws errors if op directory missing or no ops discovered
✅ Moved enhanced descriptions to separate file (scripts/operation_enhancements.json)
✅ Created docstring standard (docs/FORGE_DOCSTRING_STANDARD.md)
✅ Updated 3 operations (Resize2d, Abs, Relu) with proper docstrings to validate the approach
✅ Using clean file names (abs.md instead of forge_op_add.md)

The generator now:
1. Discovers ops from forge/forge/op/*.py (primary source of truth)
2. Parses NumPy-style docstrings including Mathematical Definition and See Also sections
3. Loads supplementary enhancements from operation_enhancements.json

To add new ops, developers just need to add a function with a proper docstring - no manual doc updates needed!

samiyadev786 commented:
@chandrasekaranpradeep @jberkowitzTT Please review this PR.

chandrasekaranpradeep reviewed (changes requested):
@samiyadev786 Thanks for addressing all the comments.  It looks we are pretty close to have these feature. Just few nice things to have:

1. **Consolidate documentation generation instructions**
   - Please move the documentation generation script usage instructions from `scripts/README_ops_docs.md` into `docs/src/tools.md` to centralize all tool documentation in one place.

2. **Clean up stale operation documentation files**
   - Currently, when operations are removed from `forge/forge/op/*.py` files (e.g., deleting a function definition like `Add` or removing an entire file like `eltwise_binary.py`), the corresponding markdown files in `docs/src/operations/` (e.g., `add.md`, `sub.md`, `mul.md`) are not automatically removed. These stale files persist even after the source operations are deleted, which can lead to outdated documentation being displayed. The documentation generation script should clean up files for operations that no longer exist in the source code.

3. **Remove legacy README file**
   - Please remove `docs/src/operations/README.md` if it exists, as it appears to be a manually maintained file from previous commits and conflicts with the auto-generated documentation structure.

4. **Extend operation enhancements support**
   - Currently, `scripts/operation_enhancements.json` only supports `mathematical_definition` and `related_operations`. It would be helpful to extend it to support additional enhancement types such as:
     - Parameter/operand descriptions (to override or supplement docstring parameter descriptions)
     - Operation descriptions (to provide enhanced overview text)
     - Other operation-specific metadata that may need manual curation

samiyadev786 commented:
> @samiyadev786 Thanks for addressing all the comments. It looks we are pretty close to have these feature. Just few nice things to have:
> 
> 1. **Consolidate documentation generation instructions**
>    
>    * Please move the documentation generation script usage instructions from `scripts/README_ops_docs.md` into `docs/src/tools.md` to centralize all tool documentation in one place.
> 2. **Clean up stale operation documentation files**
>    
>    * Currently, when operations are removed from `forge/forge/op/*.py` files (e.g., deleting a function definition like `Add` or removing an entire file like `eltwise_binary.py`), the corresponding markdown files in `docs/src/operations/` (e.g., `add.md`, `sub.md`, `mul.md`) are not automatically removed. These stale files persist even after the source operations are deleted, which can lead to outdated documentation being displayed. The documentation generation script should clean up files for operations that no longer exist in the source code.
> 3. **Remove legacy README file**
>    
>    * Please remove `docs/src/operations/README.md` if it exists, as it appears to be a manually maintained file from previous commits and conflicts with the auto-generated documentation structure.
> 4. **Extend operation enhancements support**
>    
>    * Currently, `scripts/operation_enhancements.json` only supports `mathematical_definition` and `related_operations`. It would be helpful to extend it to support additional enhancement types such as:
>      
>      * Parameter/operand descriptions (to override or supplement docstring parameter descriptions)
>      * Operation descriptions (to provide enhanced overview text)
>      * Other operation-specific metadata that may need manual curation
> 5. **Make script paths configurable via command-line arguments**
>    
>    * Currently, the paths are hardcoded in `scripts/discover_operations.py` and `scripts/generate_ops_docs.py`. It would be beneficial to add Python command-line arguments (using `argparse`) to allow users to specify:
>      
>      * The operations source directory path (currently hardcoded as `forge/forge/op/`)
>      * The output directory path for operation markdown files (currently hardcoded as `docs/src/operations/`)
>      * The output file path for the operations index page (currently hardcoded as `docs/src/operations.md`)
>      * etc.
>    * We can have default values with the hardcoded paths to maintain backward compatibility. This would make the scripts more flexible and easier to modify paths/other configurations in the future.



Hi @chandrasekaranpradeep, I've addressed all your feedback. Here's what's been implemented:

1. **Consolidated docs** - Moved documentation instructions from `scripts/README_ops_docs.md` to `docs/src/tools.md`

2. **Stale file cleanup** - Added automatic removal of orphaned markdown files when operations are deleted from source

3. **Removed legacy README** - Deleted `docs/src/operations/README.md`

4. **Extended enhancements support** - `operation_enhancements.json` now supports `description` and `parameters` overrides in addition to `mathematical_definition` and `related_operations`

5. **Configurable CLI args** - Added argparse support for `--op-dir`, `--output-dir`, `--index-file`, `--enhancements`, and `--no-cleanup` flags

samiyadev786 commented:
@chandrasekaranpradeep Please review this PR.

chandrasekaranpradeep commented:
> 
> @chandrasekaranpradeep @nvukobratTT
> 
> **`test_alexnet_onnx` - HuggingFace Authentication Error**
> 
> ```
> ConnectionError: Couldn't reach https://huggingface.co/datasets/ILSVRC/imagenet-1k/resolve/main/dataset_infos.json
> (Unauthorized for URL ... Please use the parameter `token=True` after logging in with `huggingface-cli login`)
> ```
> 
> This fails because:
> 
> * The `ILSVRC/imagenet-1k` dataset is a **gated dataset** on HuggingFace (requires access approval)
> * The CI runner doesn't have a valid HuggingFace token configured for this dataset
> * This is a **CI infrastructure/configuration issue**, not a code issue
> 
> > The CI failures in `test-models / run-tests pr_models_regression` are not related to this PR. The only actual test failure is `test_alexnet_onnx`, which fails due to a HuggingFace authentication error when trying to access the gated `ILSVRC/imagenet-1k` dataset - this is a CI infrastructure issue, not related to my documentation changes.

@samiyadev786  Thanks for sharing the failure. As you said, it is gated dataset but we have set valid HF_TOKEN in test-sub.yml workflow, so I don't think the issue was with token. In the other PR ci pipelines, we didn't face this issue and at the same I don't think the issue was caused by our code changes because I have cherry picked your commits and triggered the workflow in my custom branch, all the test are passing without above issue - https://github.com/tenstorrent/tt-forge-onnx/actions/runs/21948860177

No worries, will find out and resolve the issue

samiyadev786 commented:
> > @chandrasekaranpradeep @nvukobratTT
> > **`test_alexnet_onnx` - HuggingFace Authentication Error**
> > ```
> > ConnectionError: Couldn't reach https://huggingface.co/datasets/ILSVRC/imagenet-1k/resolve/main/dataset_infos.json
> > (Unauthorized for URL ... Please use the parameter `token=True` after logging in with `huggingface-cli login`)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This fails because:
> > 
> > * The `ILSVRC/imagenet-1k` dataset is a **gated dataset** on HuggingFace (requires access approval)
> > * The CI runner doesn't have a valid HuggingFace token configured for this dataset
> > * This is a **CI infrastructure/configuration issue**, not a code issue
> > 
> > > The CI failures in `test-models / run-tests pr_models_regression` are not related to this PR. The only actual test failure is `test_alexnet_onnx`, which fails due to a HuggingFace authentication error when trying to access the gated `ILSVRC/imagenet-1k` dataset - this is a CI infrastructure issue, not related to my documentation changes.
> 
> @samiyadev786 Thanks for sharing the failure. As you said, it is gated dataset but we have set valid HF_TOKEN in test-sub.yml workflow, so I don't think the issue was with token. In the other PR ci pipelines, we didn't face this issue and at the same I don't think the issue was caused by our code changes because I have cherry picked your commits and triggered the workflow in my custom branch, all the test are passing without above issue - https://github.com/tenstorrent/tt-forge-onnx/actions/runs/21948860177
> 
> No worries, will find out and resolve the issue

@chandrasekaranpradeep Thanks
